src/lib.rs

```diff
@@ -112,3 +115,60 @@ impl PrivilegeLevel {
         }
     }
 }
+
+/// A wrapper that can be used to safely create one mutable reference `&'static mut T` from a static variable.
+///
+/// `Singleton` is safe because it ensures that it only ever gives out one reference.
+///
+/// ``Singleton<T>` is a safe alternative to `static mut` or a static `UnsafeCell<T>`.
+#[derive(Debug)]
+pub struct Singleton<T> {
+    used: AtomicBool,
+    value: UnsafeCell<T>,
+}
+
+impl<T> Singleton<T> {
+    /// Construct a new singleton.
+    pub const fn new(value: T) -> Self {
+        Self {
+            used: AtomicBool::new(false),
+            value: UnsafeCell::new(value),
+        }
+    }
+
+    /// Try to acquire a mutable reference to the wrapped value.
+    /// This will only succeed the first time the function is
+    /// called and fail on all following calls.
+    ///
+    /// ```
+    /// use x86_64::Singleton;
+    ///
+    /// static FOO: Singleton<i32> = Singleton::new(0);
+    ///
+    /// // Call `try_get_mut` for the first time and get a reference.
+    /// let first: &'static mut i32 = FOO.try_get_mut().unwrap();
+    /// assert_eq!(first, &0);
+    ///
+    /// // Calling `try_get_mut` again will return `None`.
+    /// assert_eq!(FOO.try_get_mut(), None);
+    /// ```
+    pub fn try_get_mut(&self) -> Option<&mut T> {
+        match self
+            .used
+            .compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed)
+        {
+            Ok(_) => Some(unsafe {
+                // SAFETY: no reference has been given out yet and we won't give out another.
+                &mut *self.value.get()
+            }),
+            Err(_) => None,
+        }
+    }
+}
+
+// SAFETY: Sharing a `Singleton<T>` between threads is safe regardless of whether `T` is `Sync`
+// because we only expose the inner value once to one thread.
+unsafe impl<T> Sync for Singleton<T> {}
+
+// SAFETY: It's safe to send a `Singleton<T>` to another thread if it's safe to send `T`.
+unsafe impl<T: Send> Send for Singleton<T> {}
```

Comment:
I think we still need the `T: Send` / `T: Sync` bounds here because the value might be constructed on one thread and used on another.

Reply:
Rust's documentation for `Send` is a bit misleading in my opinion, it states that a type implements `Send` when it is safe to send the object to another thread. My understanding from this was, that it only needs be safe to transfer ownership to another thread, however it seems that "sending the object to another thread" also refers to sending a unique reference to another thread. I updated the `Sync` impl accordingly by adding a `T: Send` bound.

I still think that we don't need a `Sync` bound though. [`Mutex`](https://doc.rust-lang.org/std/sync/struct.Mutex.html#impl-Sync) also has the same bounds requiring only `T: Send` for both `Sync` and `Send`.

Reply:
Good idea to compare the bounds with `Mutex`. I think that you're right that only a `Send` bound is needed since we never share a reference between threads. A `&'static mut` reference is basically ownership because it gives unique access to the value for the remainder of its life.